tools/prow-auto-bumper/parser.go

```diff
@@ -0,0 +1,178 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// prow-auto-bumper finds stable Prow components version used by k8s,
+// and creates PRs updating them in knative/test-infra
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"math"
+	"strings"
+	"time"
+
+	"github.com/knative/test-infra/shared/ghutil"
+)
+
+// Tags could be in the form of: v[YYYYMMDD]-[GIT_HASH](-[VARIANT_PART]),
+// separate it to `v[YYYYMMDD]-[GIT_HASH]` and `[VARIANT_PART]`
+func deconstructTag(in string) (string, string) {
+	dateCommit := in
+	var variant string
+	parts := strings.Split(in, "-")
+	if len(parts) > 2 {
+		variant = strings.Join(parts[2:], "-")
+	}
+	if len(parts) > 1 {
+		dateCommit = fmt.Sprintf("%s-%s", parts[0], parts[1])
+	}
+	return dateCommit, variant
+}
+
+// get key with highest value
+func getDominantKey(m map[string]int) string {
+	var res string
+	for key, v := range m {
+		if "" == res || v > m[res] {
+			res = key
+		}
+	}
+	return res
+}
+
+func (pv *PRVersions) getDominantVersions() versions {
+	if nil != pv.dominantVersions {
+		return *pv.dominantVersions
+	}
+
+	cOld := make(map[string]int)
+	cNew := make(map[string]int)
+	for _, vss := range pv.images {
+		for _, vs := range vss {
+			normOldTag, _ := deconstructTag(vs.oldVersion)
+			normNewTag, _ := deconstructTag(vs.newVersion)
+			cOld[normOldTag]++
+			cNew[normNewTag]++
+		}
+	}
+
+	pv.dominantVersions = &versions{
+		oldVersion: getDominantKey(cOld),
+		newVersion: getDominantKey(cNew),
+	}
+
+	return *pv.dominantVersions
+}
+
+// parse changelist, find all version changes, and store them in image name: versions map
+func (pv *PRVersions) parseChangelist(gcw *GHClientWrapper, gi gitInfo) error {
+	fs, err := gcw.ListFiles(gi.org, gi.repo, *pv.PR.Number)
+	if nil != err {
+		return err
+	}
+	for _, f := range fs {
+		if nil == f.Patch {
+			continue
+		}
+		minuses := imageMinusRegexp.FindAllStringSubmatch(*f.Patch, -1)
+		for _, minus := range minuses {
+			iv := pv.getIndex(minus[imageImagePart], minus[imageTagPart])
+			pv.images[minus[imageImagePart]][iv].oldVersion = minus[imageTagPart]
+		}
+
+		pluses := imagePlusRegexp.FindAllStringSubmatch(*f.Patch, -1)
+		for _, plus := range pluses {
+			iv := pv.getIndex(plus[imageImagePart], plus[imageTagPart])
+			pv.images[plus[imageImagePart]][iv].newVersion = plus[imageTagPart]
+		}
+	}
+
+	return nil
+}
+
+// Query all PRs from "k8s-ci-robot:autobump", find PR roughly 7 days old and was not reverted later.
+// Only return error if it's github related
+func getBestVersion(gcw *GHClientWrapper, gi gitInfo) (*PRVersions, error) {
+	visited := make(map[string]PRVersions)
+	var bestPv *PRVersions
+	var overallErr error
+	var bestDelta float64 = maxDelta + 1
```

Comment:
I was trying to find where `maxDelta` is, then I realized it was in `main.go`. I think it would clearer to put these constants in a separate file like `constants.go` or `config.go`.

Reply:
good idea, moved all these to config.go